bridge/jira/export.go

```diff
@@ -0,0 +1,449 @@
+			// "issueType" field is always required. 10001 is "story" which I'm
+			// pretty sure is standard/default on all JIRA instances.
+			fields["issueType"] = "10001"
+		}
+		bugIDField, hasConf := self.conf[keyCreateGitBug]
+		if hasConf {
+			// If the git configuration also indicates it, we can assign the git-bug
+			// id to a custom field to assist in integrations
+			fields[bugIDField] = b.Id().String()
+		}
+
+		// create bug
+		result, err := client.CreateIssue(
+			self.project.ID, createOp.Title, createOp.Message, fields)
+		if err != nil {
+			err := errors.Wrap(err, "exporting jira issue")
+			out <- core.NewExportError(err, b.Id())
+			return
+		}
+
+		id := result.ID
+		out <- core.NewExportBug(b.Id())
+		// mark bug creation operation as exported
+		err = markOperationAsExported(
+			b, createOp.Id(), id, self.project.Key, time.Time{})
+		if err != nil {
+			err := errors.Wrap(err, "marking operation as exported")
+			out <- core.NewExportError(err, b.Id())
+			return
+		}
+
+		// commit operation to avoid creating multiple issues with multiple pushes
+		err = b.CommitAsNeeded()
+		if err != nil {
+			err := errors.Wrap(err, "bug commit")
+			out <- core.NewExportError(err, b.Id())
+			return
+		}
+
+		// cache bug jira ID
+		bugJiraID = id
+	}
+
+	// cache operation jira id
+	self.cachedOperationIDs[createOp.Id()] = bugJiraID
+
+	// lookup the mapping from git-bug "status" to JIRA "status" id
+	statusMap := getStatusMap(self.conf)
+
+	for _, op := range snapshot.Operations[1:] {
+		// ignore SetMetadata operations
+		if _, ok := op.(*bug.SetMetadataOperation); ok {
+			continue
+		}
+
+		// ignore operations already existing in jira (due to import or export)
+		// cache the ID of already exported or imported issues and events from
+		// Jira
+		if id, ok := op.GetMetadata(keyJiraID); ok {
+			self.cachedOperationIDs[op.Id()] = id
+			out <- core.NewExportNothing(op.Id(), "already exported operation")
+			continue
+		}
+
+		opAuthor := op.GetAuthor()
+		client, err := self.getIdentityClient(&ctx, opAuthor.Id())
+		if err != nil {
+			out <- core.NewExportNothing(
+				op.Id(), fmt.Sprintf(
+					"missing operation author credentials for user %.8s",
+					author.Id().String()))
+			continue
+		}
+
+		var id string
+		var exportTime time.Time
+		switch opr := op.(type) {
+		case *bug.AddCommentOperation:
+			comment, err := client.AddComment(bugJiraID, opr.Message)
+			if err != nil {
+				err := errors.Wrap(err, "adding comment")
+				out <- core.NewExportError(err, b.Id())
+				return
+			}
+			id = comment.ID
+			out <- core.NewExportComment(op.Id())
+
+			// cache comment id
+			self.cachedOperationIDs[op.Id()] = id
+
+		case *bug.EditCommentOperation:
+			if opr.Target == createOp.Id() {
+				// An EditCommentOpreation with the Target set to the create operation
+				// encodes a modification to the long-description/summary.
+				exportTime, err = client.UpdateIssueBody(bugJiraID, opr.Message)
```

Comment:
there is probably a way to pass the context up to the actual request to have a proper cancellation

Reply:
Good point.